### DISCUSSION TITLE
Direct Message Passing along Edges

### DISCUSSION THREAD

**Discussion by fredhastedt** [pyg-team/pytorch_geometric#7587] [Q&A] [answered]
Hi there, 

I am new to pyg and was wondering about the following use-case for a specific MessagePassing Network implementation:
Instead of updating the node features during each layer (aggregation/update) I would just like to return the message along each edge within the graph. This would roughly translate to: m_{j,i}^t = f(x_i, e_{i,j}, m_{j,i}^{t-1}). 

An example of this method can be found in the following paper (Appendix A): 
_Learning Graph Models for Retrosynthesis Prediction_ ,[https://arxiv.org/abs/2006.07038](url)

If this is possible, does this also go well with the Explainer algorithms? 
Thank you for your help!

**Comment by rusty1s** [comment]
I think you can directly do this via
```python
x_j = x[edge_index[0]]
x_i = x[edge_index[1]]

f(x_i, edge_attr, x_j)
```

**Reply by fredhastedt** [reply]
Hi Matthias, 

So I guess the easiest way to use some Explainer algorithm for that model would be to subclass the MessagePassing baseclass and then to simply call self.message(*args) within the forward pass of my model. Is that correct? 
Thanks!

**Reply by fredhastedt** [reply]
I guess the main concern/confusion that I have is how to incorporate the existing functionality for the explainer algorithm with a model that doesn't necessarily use the propagate method within the forward pass.

**Reply by rusty1s** [reply, marked as answer]
Mh, I see. I missed the part where you want to use the explainer :(

What you can probably do is use `MessagePassing` but overwrite the `aggregate` function, e.g.:
```python
def aggregate(self, inputs):
    return inputs
```
This way, the explainer still manages to weight the importance of different messages.

**Reply by fredhastedt** [reply]
Thanks for this, makes a lot of sense and pretty much answers my question. 

If I may add another one - might have been answered already (please excuse me if it has). 
For getting the neighbouring nodes during the message function, is there a special variable that I can pass to the function e.g. I’ve seen the index variable passed directly to the softmax in the HEATConv model? 
Furthermore, when using the explainer for this type of model, would this pertain to the take of edge explanation? (Since I will have a tensor returned of shape (E, hidden_dim)? 

Thank you for taking the time; I really appreciate your help. Also, this library is really superb for geometric deep learning!

**Reply by rusty1s** [reply]
Yes, accessing `index` in `aggregate` or `message` gives you the destination node index for every edge.

> Furthermore, when using the explainer for this type of model, would this pertain to the take of edge explanation? (Since I will have a tensor returned of shape (E, hidden_dim)?

I don't see any issues with this.
